packages/typed-config/src/index.ts

```diff
@@ -0,0 +1,317 @@
+declare global {
+  // a namespace makes it easy to group all of the types together
+  // so that they don't pollute the global namespace
+  // eslint-disable-next-line @typescript-eslint/no-namespace
+  namespace ESLintConfig {
+    export type RuleLevel = 0 | 1 | 2 | 'off' | 'warn' | 'error';
+    export type RuleLevelAndNoOptions = RuleLevel | [RuleLevel];
+    export type RuleLevelAndUnknownOptions =
```

Comment:
Also add a rule-type for known options so that rule-authors can add their own definition:

```ts
export type RuleLevelAndKnownOptions<TRuleName extends keyof RuleStore> = RuleStore[TRuleName];
interface RuleStore
{
    "space": [
        RuleLevel,
        ...Array<(
            "tab" |
            number |
            {
                SwitchCase: number
            })>
    ];
}
```